ChangeLog entry:

2005-09-26  Mark D. Baushke  <mdb@gnu.org>

	* ci.c (mainProg): Add commitid to delta records using the
	same basic format construction as is used by CVS and CVSNT.
	* rcsbase.h (commitidsize): Room for 64bit pid, 32bit time,
	16bit random rendered as hex plus one NUL byte round up to 32.
	(struct hshentry): Add new commitid field.
	* rcsgen.c (putdelta): Preserve old commitid entries.
	* rcssyn.c (Kcommitid): New global constant keyword.
	(getdelta): Add optional parsing for it.
	* rlog.c (putadelta): Print it out.

Index:src/ci.c
--- src/ci.c~	1995-06-16 06:19:24.000000000 +0000
+++ src/ci.c	2005-09-26 16:47:38.435505000 +0000
@@ -285,6 +285,7 @@ mainProg(ciId, "ci", "$Id: ci.c,v 5.30 1
 	char olddate[datesize];
 	char newdatebuf[datesize + zonelenmax];
 	char targetdatebuf[datesize + zonelenmax];
+	char commitid[commitidsize];
 	char *a, **newargv, *textfile;
 	char const *author, *krev, *rev, *state;
 	char const *diffname, *expname;
@@ -309,6 +310,9 @@ mainProg(ciId, "ci", "$Id: ci.c,v 5.30 1
 	suffixes = X_DEFAULT;
 	nextassoc = &assoclst;
 
+	snprintf(commitid, sizeof(commitid), "%x%08lx%04x",
+		 (int)getpid(), (long)time (NULL), rand()&0xFFFF);
+
 	argc = getRCSINIT(argc, argv, &newargv);
 	argv = newargv;
 	while (a = *++argv,  0<--argc && *a++=='-') {
@@ -532,6 +536,8 @@ mainProg(ciId, "ci", "$Id: ci.c,v 5.30 1
 	newdelta.name = 0;
 	clear_buf(&newdelta.ig);
 	clear_buf(&newdelta.igtext);
+	/* set commitid */
+	newdelta.commitid=commitid;
 	/* set author */
 	if (author)
 		newdelta.author=author;     /* set author given by -w         */
Index:src/rcsbase.h
--- src/rcsbase.h~	1995-06-16 06:19:24.000000000 +0000
+++ src/rcsbase.h	2005-09-26 16:48:34.726505000 +0000
@@ -222,6 +222,7 @@ Report problems and direct all questions
                               /* 1 sets the default locking to strict;      */
                               /* used in production environments.           */
 
+#define commitidsize	   32 /* pid + time + random bits                   */
 #define yearlength	   16 /* (good through AD 9,999,999,999,999,999)    */
 #define datesize (yearlength+16)	/* size of output of time2date */
 #define RCSTMPPREFIX '_' /* prefix for temp files in working dir  */
@@ -358,6 +359,7 @@ struct hshentry {
 	char const	  * lockedby; /* who locks the revision		    */
 	char const	  * state;    /* state of revision (Exp by default) */
 	char const	  * name;     /* name (if any) by which retrieved   */
+	char const        * commitid; /* text string to associate commits   */
 	struct cbuf	    log;      /* log message requested at checkin   */
         struct branchhead * branches; /* list of first revisions on branches*/
 	struct cbuf	    ig;	      /* ignored phrases in admin part	    */
@@ -662,6 +664,7 @@ extern int               TotalDeltas;
 extern char const *const expand_names[];
 extern char const
 	Kaccess[], Kauthor[], Kbranch[], Kcomment[],
+	Kcommitid[],
 	Kdate[], Kdesc[], Kexpand[], Khead[], Klocks[], Klog[],
 	Knext[], Kstate[], Kstrict[], Ksymbols[], Ktext[];
 void unexpected_EOF P((void)) exiting;
Index:src/rcsgen.c
--- src/rcsgen.c~	1995-06-16 06:19:24.000000000 +0000
+++ src/rcsgen.c	2005-09-23 18:09:11.408920000 +0000
@@ -547,6 +547,9 @@ putdelta(node, fout)
 
 	aprintf(fout, ";\n%s\t%s;\n", Knext, node->next?node->next->num:"");
 	awrite(node->ig.string, node->ig.size, fout);
+
+	if (node->commitid)
+		aprintf(fout, "%s\t%s;\n", Kcommitid, node->commitid);
 }
 
 
Index:src/rcssyn.c
--- src/rcssyn.c~	1995-06-16 06:19:24.000000000 +0000
+++ src/rcssyn.c	2005-09-23 18:01:47.652920000 +0000
@@ -171,6 +171,7 @@ char const
 	Kauthor[]   = "author",
 	Kbranch[]   = "branch",
 	Kcomment[]  = "comment",
+	Kcommitid[] = "commitid",
 	Kdate[]     = "date",
 	Kdesc[]     = "desc",
 	Kexpand[]   = "expand",
@@ -433,6 +434,13 @@ getdelta()
 	Delta->lockedby = 0;
 	Delta->log.string = 0;
 	Delta->selector = true;
+
+	if (getkeyopt(Kcommitid)) {
+		Delta->commitid = NextString;
+		nextlex();
+		getsemi(Kcommitid);
+        }
+
 	Delta->ig = getphrases(Kdesc);
         TotalDeltas++;
         return (true);
Index:src/rlog.c
--- src/rlog.c~	1995-06-16 06:19:24.000000000 +0000
+++ src/rlog.c	2005-09-26 17:23:55.257504000 +0000
@@ -591,6 +591,10 @@ putadelta(node,editscript,trunk)
 	      aprintf(out, insDelFormat,
                              editscript->insertlns, editscript->deletelns);
 
+	if ( node->commitid )
+	   aprintf(out, "%s commitid: %s", (editscript) ? ";" : "",
+		   node->commitid);
+
         newbranch = node->branches;
         if ( newbranch ) {
 	   bufautobegin(&branchnum);
